pkg/agent/controller/l7visibility/l7visibility_controller.go

```diff
@@ -0,0 +1,404 @@
+	}
+	<-stopCh
+}
+
+func (l7c *L7VisibilityController) worker() {
+	for l7c.processNextWorkItem() {
+	}
+}
+
+func (l7c *L7VisibilityController) processNextWorkItem() bool {
+	obj, quit := l7c.queue.Get()
+	if quit {
+		return false
+	}
+	defer l7c.queue.Done(obj)
+
+	if key, ok := obj.(string); !ok {
+		// As the item in the work queue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen.
+		l7c.queue.Forget(key)
+		klog.ErrorS(nil, "Expected string in work queue but got", "key", obj)
+		return true
+	} else if err := l7c.syncPod(key); err == nil {
+		// If no error occurs we Forget this item, so it does not get queued again until
+		// another change happens.
+		l7c.queue.Forget(key)
+	} else if err == errInvalidAnnotation || err == errPodInterfaceNotFound {
+		// Handle errors
+		l7c.queue.Forget(key)
+	} else {
+		// Put the item back on the work queue to handle any transient errors.
+		l7c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Syncing Pod object for L7Visibility failed, requeue", "Pod", key)
+	}
+	return true
+}
+
+func (l7c *L7VisibilityController) addPod(obj interface{}) {
+	pod := obj.(*v1.Pod)
+	if !isValidPod(pod) {
+		return
+	}
+	_, ok := pod.Annotations[types.L7VisibilityAnnotationKey]
+	if !ok {
+		if !l7c.namespaceAnnotationExists(pod) {
+			return
+		}
+	}
+	klog.V(2).InfoS("Processing Pod ADD event", "Pod", klog.KObj(pod))
+	podNN := k8s.NamespacedName(pod.Namespace, pod.Name)
+	l7c.queue.Add(podNN)
+}
+
+func (l7c *L7VisibilityController) updatePod(oldObj interface{}, obj interface{}) {
+	oldPod := oldObj.(*v1.Pod)
+	updatedPod := obj.(*v1.Pod)
+	if !isValidPod(updatedPod) {
+		return
+	}
+	oldAnnotation, _ := oldPod.GetAnnotations()[types.L7VisibilityAnnotationKey]
+	updatedAnnotation, updatedAnnotationOk := updatedPod.GetAnnotations()[types.L7VisibilityAnnotationKey]
+	if oldPod.Status.PodIP != updatedPod.Status.PodIP {
+		if !updatedAnnotationOk && !l7c.namespaceAnnotationExists(updatedPod) {
+			return
+		}
+	} else if oldPod.Status.PodIP == updatedPod.Status.PodIP {
+		if oldAnnotation != updatedAnnotation && updatedAnnotationOk {
+			if l7c.namespaceAnnotationExists(updatedPod) {
+				// new need to add to queue, NS event took care of the Pod
+				return
+			}
+		}
+	}
+
+	klog.V(2).InfoS("Processing Pod UPDATE event", "Pod", klog.KObj(updatedPod))
+	podNN := k8s.NamespacedName(updatedPod.Namespace, updatedPod.Name)
+	l7c.queue.Add(podNN)
+}
+
+func (l7c *L7VisibilityController) deletePod(obj interface{}) {
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(nil, "Received unexpected object", "object", obj)
+			return
+		}
+		pod, ok = deletedState.Obj.(*v1.Pod)
+		if !ok {
+			klog.ErrorS(nil, "DeletedFinalStateUnknown contains non-Pod object", "object", deletedState.Obj)
+			return
+		}
+	}
+	if !isValidPod(pod) {
+		return
+	}
```

Comment:
Is it reasonable to have this check here? I don't recall if it's possible for the `pod.Status.PodIP` field to be empty when we receive the DELETE event. I feel like it would be safer not to assume that `pod.Status.PodIP` is set when handling deletion.

Reply:
if a pod is in Pending state and is deleted will it not come to the delete function ?

Reply:
It is possible. But it is also possible (IMO) to have the following events:
- ADD (no IP yet)
- UPDATE (IP is present)
- DELETE (no IP any more)

just like you can have:
- ADD (no IP yet)
- UPDATE (IP is present)
- DELETE (IP is still present)

In the first sequence above, you still want to be able to clean up resources in your controller as needed (e.g. update the map). So I don't think we should skip the DELETE event if the IP is not present, unless you can find a strong documented guarantee that the first sequence is not valid.

Reply:
This may be helpful: https://pkg.go.dev/k8s.io/client-go/tools/cache#ResourceEventHandler, though it doesn't talk about the continuity of events.
> * OnAdd is called when an object is added.
> * OnUpdate is called when an object is modified. Note that oldObj is the last known state of the object-- it is possible that several changes were combined together, so you can't use this to see every single change. OnUpdate is also called when a re-list happens, and it will get called even if nothing changed. This is useful for periodically evaluating or syncing something.
> * OnDelete will get the final state of the item if it is known, otherwise it will get an object of type DeletedFinalStateUnknown. This can happen if the watch is closed and misses the delete event and we don't notice the deletion until the subsequent re-list.

From my understanding, the events are generated based on the states of cache. If the Pod IP was ever removed from the object, there would be a UPDATE event indicating the change, otherwise this update was missed by the cache (due to disconnection), and when the cache find the Pod is missing in the new list (got via relisting), it should generate a delete event with the last known state (which contains the Pod IP). 

One case I want to mention is there could be ADD event with Pod IP present. This happens when a Pod already exists when agent starts, the informer will only generate an ADD event as it's state based, not event based.

Reply:
thanks @tnqn 

If the following sequence is at all possible (maybe unlikely):
1) UPDATE to remove the IP
2) DELETE (no Pod IP in last know state - see previous UPDATE)

Then I feel like the current code is not valid:

```go
func (l7c *L7VisibilityController) addPod(obj interface{}) {
	pod := obj.(*v1.Pod)
	if !isValidPod(pod) {
		return
	}
	// ...
}

func (l7c *L7VisibilityController) updatePod(oldObj interface{}, obj interface{}) {
	oldPod := oldObj.(*v1.Pod)
	updatedPod := obj.(*v1.Pod)
	if !isValidPod(updatedPod) {
		return
	}
	// ...
}

func (l7c *L7VisibilityController) deletePod(obj interface{}) {
	pod, ok := obj.(*v1.Pod)
	if !ok {
		// ...
	}
	if !isValidPod(pod) {
		return
	}
	// ...
}
```

The `!isValidPod(updatedPod)` gate in both `updatePod` & `deletePod` would mean that the Antrea state is not cleaned up for the Pod, unless I am missing something.

That's why I'm advocating for removing the gate for `deletePod`. This is similar to what we do for the NPL controller for example. This seems like the simplest way to guarantee proper clean up. Although maybe it's possible to achieve the same thing by changing the logic in `updatePod`.